Họ tên: Nguyễn Bửu Thạch

MSSV: 20120576

# HW0: Làm quen với CUDA

Với các GPU tương đối mới thì để biên dịch chỉ cần dùng câu lệnh: \
`nvcc tên-file.cu -o tên-file-chạy`

Nhưng trên Colab mình thường lấy được GPU khá cũ là Tesla K80 với compute capability (phiên bản phần cứng) là 3.7; để biên dịch đúng với GPU khá cũ này thì bạn cần dùng câu lệnh: \
`nvcc -arch=sm_37 tên-file.cu -o tên-file-chạy` \
Trong đó, 37 chính là compute capability của GPU Tesla K80.

Để phòng trường hợp khi làm bài bạn lấy được GPU có compute capability x.x nhưng khi chấm bài Thầy lại lấy được GPU có compute capability khác x.x, dưới đây mình sẽ có đoạn code Python để tự động lấy 2 con số ứng với compute capability của GPU và lưu vào 2 biến `major` và `minor`:


In [ ]:
from numba import cuda
major, minor = cuda.get_current_device().compute_capability
print(f'GPU compute capability: {major}.{minor}')

Một khi đã chạy đoạn code Python ở trên, để biên dịch thì bạn sẽ dùng câu lệnh: \
`nvcc -arch=sm_{major}{minor} tên-file.cu -o tên-file-chạy`

Dưới đây, khi làm bài thì bạn có thể tùy ý thêm/xóa cell. Đừng xóa mấy cell có chữ của Thầy là được.

## Câu 1

In [1]:
!nvcc HW0_P1.cu -o HW0_P1

In [2]:
!./HW0_P1

Information GPU Card: 
1/GPU card's name: Tesla T4
2/GPU computation capabilities: 7.5
3/Maximum number of block dimensions: 1024 x 1024 x 64
4/Maximum number of grid dimensions: 2147483647 x 65535 x 65535
5/Maximum size of GPU memory: 15835660288 bytes
6/Amount of constant and share memory: 
Amount of constant memory: 65536 bytes
Amount of shared memory: 49152 bytes
7/Warp size: 32


Kết quả sau khi chạy chương trình:


Information GPU Card:\
1/GPU card's name: Tesla T4 \
2/GPU computation capabilities: 7.5 \
3/Maximum number of block dimensions: 1024 x 1024 x 64 \
4/Maximum number of grid dimensions: 2147483647 x 65535 x 65535 \
5/Maximum size of GPU memory: 15835660288 bytes \
6/Amount of constant and share memory: \
Amount of constant memory: 65536 bytes \
Amount of shared memory: 49152 bytes \
7/Warp size: 32 \

## Câu 2

In [6]:
!nvcc HW0_P2.cu -o HW0_P2

In [7]:
!./HW0_P2

Vector Size = 64
Processing time (use host): 0.010176 ms
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 1): 0.23552 ms
CORRECT for version 1
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 2): 0.02192 ms
CORRECT for version 2
Vector Size = 256
Processing time (use host): 0.00896 ms
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 1): 0.034304 ms
CORRECT for version 1
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 2): 0.024608 ms
CORRECT for version 2
Vector Size = 1024
Processing time (use host): 0.012896 ms
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 1): 0.026624 ms
CORRECT for version 1
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 2): 0.024608 ms
CORRECT for version 2
Vector Size = 4096
Processing time (use host): 0.046784 ms
GPU name: Tesla T4
GPU compute capability: 7.5
Processing time (use version 1): 0.0269

Kết quả sau khi chạy chương trình:

| Vector size | Host time | Device time (Version 1) | Device time (Version 2) |  
|-------------|-----------|-------------------------|--------------------------|  
|64|0.010176 ms|0.23552 ms|0.02192 ms|
|256|0.00896 ms|0.034304 ms|0.024608 ms|
|1024|0.012896 ms|0.026624 ms|0.024608 ms|
|4096|0.046784 ms|0.026976 ms|0.025088 ms|
|16384|0.086624 ms|0.02416 ms|0.022336 ms|
|65536|0.330528 ms|0.02672 ms|0.024864 ms|
|262144|1.27485 ms|0.036256 ms| 0.03216 ms|
|1048576|4.89677 ms|0.080224 ms|0.081184 ms|
|4194304|20.0278 ms|0.259296 ms|0.26864 ms|
|16777216|81.67 ms|0.936064 ms|1.02307 ms|





Em có tham khảo file 01-AddVector.cu và có sử dụng lại  GpuTimer, hàm addVecOnHost, hàm AddVec và int main() để làm câu 2